### PR TITLE
gh-110014: Fix _POSIX_THREADS and _POSIX_SEMAPHORES usage

### PR DESCRIPTION
**Pull request by vstinner**
* pycore_pythread.h is now the central place to make sure that _POSIX_THREADS and _POSIX_SEMAPHORES macros are defined if available.
* Make sure that pycore_pythread.h is included when _POSIX_THREADS and _POSIX_SEMAPHORES macros are tested.
* PY_TIMEOUT_MAX is now defined as a constant, since its value depends on _POSIX_THREADS, instead of being defined as a macro.
* Document the change and give hints how to fix affected code.

* Issue: gh-110014

----
:books: Documentation preview :books:: https://cpython-previews--110139.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
@encukou: "Oops", I have to add an exception to ``Tools/build/smelly.py`` to tolerate the new symbol ``PY_TIMEOUT_MAX``. Before, Python only exported symbols starting with ``Py``, ``_Py`` and ``__Py`` (macOS). Copy of my comment:

```py
# "Legacy": some old symbols are prefixed by "PY_".
EXCEPTIONS = frozenset({
    'PY_TIMEOUT_MAX',
})
```

I would prefer to prefix all public symbols with ``Py`` and only have private symbols, but at least, now we have tooling to detect when new symbols with different prefix are added ;-)

**Comment by vstinner** [comment]
Oh. I don't know if it's related or not, but I had to write PR gh-110155 to add many ``<unistd.h>``. WASM/WASI builds were badly broken :-(

File: Misc/stable_abi.toml [hunk 1]
```diff
@@ -1843,6 +1843,10 @@
 [function.PyThread_start_new_thread]
     added = '3.2'
 
+# Not mentioned in PEP 384, was implemented as a macro in Python <= 3.12
+[data.PY_TIMEOUT_MAX]
```
Comment (encukou) [inline Misc/stable_abi.toml:1847]: @vstinner Could we instead exclude this undocumented constant from the limited API?

Comment (vstinner) [reply Misc/stable_abi.toml:1847]: I checked and apparently, it's safe to remove this constant, so I created PR gh-110217 for that.

**Pull request merged**